packages/shorebird_cli/lib/src/commands/preview_command.dart

```diff
@@ -0,0 +1,186 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:mason_logger/mason_logger.dart';
+import 'package:path/path.dart' as p;
+import 'package:shorebird_cli/src/adb.dart';
+import 'package:shorebird_cli/src/bundletool.dart';
+import 'package:shorebird_cli/src/cache.dart';
+import 'package:shorebird_cli/src/code_push_client_wrapper.dart';
+import 'package:shorebird_cli/src/command.dart';
+import 'package:shorebird_cli/src/logger.dart';
+import 'package:shorebird_cli/src/shorebird_config_mixin.dart';
+import 'package:shorebird_cli/src/shorebird_validation_mixin.dart';
+import 'package:shorebird_cli/src/third_party/flutter_tools/lib/flutter_tools.dart';
+
+/// {@template preview_command}
+/// `shorebird preview` command.
+/// {@endtemplate}
+class PreviewCommand extends ShorebirdCommand
+    with ShorebirdConfigMixin, ShorebirdValidationMixin {
+  /// {@macro preview_command}
+  PreviewCommand({super.validators}) {
+    argParser
+      ..addOption(
+        'device-id',
+        help: 'The ID of the device or simulator to preview the release on.',
+      )
+      ..addOption(
+        'app-id',
+        help: 'The ID of the app to preview the release for.',
+      )
+      ..addOption(
+        'release-version',
+        help: 'The version of the release (e.g. "1.0.0").',
+      );
+  }
+
+  @override
+  String get name => 'preview';
+
+  @override
+  String get description => 'Preview a specific release on a device.';
+
+  @override
+  Future<int> run() async {
+    try {
+      await validatePreconditions(checkUserIsAuthenticated: true);
+    } on PreconditionFailedException catch (error) {
+      return error.exitCode.code;
+    }
+
+    Future<String?> promptForApp() async {
```

Comment:
Can this be defined outside of the run() function? Functions in functions disrupt the control flow and make the code a bit harder to parse.

Reply:
done ✅